### PR TITLE
Add `Accelerator.is_available()` interface requirement

### LINKED ISSUE #11818: [RFC] `Add Accelerator.is_available()` interface requirement
### Motivation

Such functionality on the Accelerator abstraction would:
- Enable automatic hardware selection without duplicating code across Trainer & individual accelerator implementations.
- Simplify the accelerator connector logic and rewrite effort: https://github.com/PyTorchLightning/pytorch-lightning/pull/11448
- Enable automatic runtime checking of hardware availability during execution
- Provide consistency with how the Trainer auto-detects the cluster environments natively supported by the framework. The corollary here is `ClusterEnvironment.detect`

https://github.com/PyTorchLightning/pytorch-lightning/blob/9e63281a4c4a62f32cad9801a23b63454f8311be/pytorch_lightning/plugins/environments/cluster_environment.py#L43-L46

https://github.com/PyTorchLightning/pytorch-lightning/blob/9e63281a4c4a62f32cad9801a23b63454f8311be/pytorch_lightning/trainer/connectors/accelerator_connector.py#L810-L812

### Pitch

```py
class Accelerator(ABC):

    @staticmethod
    @abstractmethod
    def is_available() -> bool:
        """Detect if the hardware is available."""
    
    def setup_environment(self, root_device: torch.device) -> None:
        """Setup any processes or distributed connections.
        This is called before the LightningModule/DataModule setup hook which allows the user to access the accelerator
        environment before setup is complete.
        Raises:
            RuntimeError:
                If corresponding hardware is not found.
        """
        if not self.is_available():
            raise RuntimeError(f"{self.__class__.__qualname__} is not configured to run on this hardware.")
```

```py
class CPUAccelerator(Accelerator):
    @staticmethod
    def is_available() -> bool:
        """CPU is always available for execution."""
        return True

```

```py
class GPUAccelerator(Accelerator):
   @staticmethod
    def is_available() -> bool:
        return torch.cuda.is_available() and torch.cuda.device_count() > 0 
```

and so on

See a more-detailed implementation here: https://github.com/PyTorchLightning/pytorch-lightning/pull/11797 for what this looks like in practice

To support `Trainer(accelerator="auto")` this is what the logic simplifies to:

```py
for acc_cls in (GPUAccelerator, TPUAccelerator, IPUAccelerator, CPUAccelerator):
    if acc_cls.is_available():
        return acc_cls()
return CPUAccelerator() # fallback to CPU
```

this could be even further simplified if we offered an AcceleratorRegistry, such that the Trainer/AcceleratorConnector didn't need to hardcode the list of accelerators to detect:
```py
for acc_cls in AcceleratorRegistry.impls:
    if acc_cls.is_available():
        return acc_cls()
return CPUAccelerator() # fallback to CPU
```

### Alternatives

Some other alternatives exist here: 
https://github.com/PyTorchLightning/pytorch-lightning/pull/11799
https://github.com/PyTorchLightning/pytorch-lightning/pull/11798

Issues with these approaches:
- Also breaking changes: simply instantiating the accelerator could raise a runtime error if the device isn't available.
- The bigger issue to me is that it does not ease support for `Trainer(accelerator="auto")`. The accelerator connector needs to hardcode & re-implement each of the device checks to determine which Accelerator to even instantiate.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @tchaton @justusschock @awaelchli @akihironitta @rohitgr7

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

- Enables automatic hardware selection without duplicating code across Trainer & individual accelerator implementations. This can be further enhanced by the addition of an AcceleratorRegistry. The accelerator connector could iterate through all of the registered accelerators, call `acc_cls.is_available()` and determine the hardware immediately.

This aims to simplify the accelerator connector logic and rewrite effort in #11448
This would move the hardcoded, duplicated assertion logic from Trainer constructor checks to a single runtime check at `trainer.fit/validate/test/predict` calls

Given discussion on this PR, we can decide where the assertion on device availability should happen. Either in the accelerator init, setup_environment, or left up to individual accelerators to decide

Fixes #11818

### Does your PR introduce any breaking changes? If yes, please list them.
Yes, this now:
- raises a TypeError if a custom accelerator has not implemented the new abstract method
- raises a RuntimeError if the configured hardware is not available during Trainer execution

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [n/a] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/gpu.py
Comment (carmocca) [inline pytorch_lightning/accelerators/gpu.py:41]:
```suggestion
                If torch.cuda isn't available.
                If no CUDA devices are found.
```

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/accelerators/gpu.py
Comment (rohitgr7) [inline pytorch_lightning/accelerators/gpu.py:47]: quick ques: why not inside accelerator init?

Comment (four4fish) [reply pytorch_lightning/accelerators/gpu.py:47]: User should be able to create a accelerator object in any environment, we only want to check device availability when call trainer.fit/test/eval, and setup_evn is the first hook called in trainer._run

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/accelerators/gpu.py
Comment (carmocca) [reply pytorch_lightning/accelerators/gpu.py:47]: I would also prefer the `__init__`, it's generally where strategies run their checks:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/ipu.py#L98-L102

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/deepspeed.py#L263-L266

Comment (rohitgr7) [reply pytorch_lightning/accelerators/gpu.py:47]: > User should be able to create a accelerator object in any environment, we only want to check device availability when call trainer.fit/test/eval, and setup_evn is the first hook called in trainer._run

quick ques: can you explain more on this? do you mean creating obj in non-GPU env and using it in GPU env? is that possible?

Comment (four4fish) [reply pytorch_lightning/accelerators/gpu.py:47]: @rohitgr7 I think it's possible in production/complex system, for example: User needs to create obj in a config script, and then send the config file to a scheduler to trigger/schedule the training job.

Comment (four4fish) [reply pytorch_lightning/accelerators/gpu.py:47]: >But to the best of my knowledge, the config files define the structure of the object constructor but don't actually initialize them until the job is executed.

Agree. :) 
I saw our internal users/team have config wrappers on top of lightning (for reusability), where defined accelerator obj. I'm not sure whether it will fail if we add check in `__init__`, but feel better to be safe.

Comment (ananthsub) [reply pytorch_lightning/accelerators/gpu.py:46]: I can't recall any instance where the GPU accelerator is specified on a host which does not have GPU devices. In almost every instance I've seen, the entire training script is packaged and run on a GPU host.

I moved the assertion to happen back during `__init__` to fail faster

File: pytorch_lightning/accelerators/gpu.py
Comment (rohitgr7) [inline pytorch_lightning/accelerators/gpu.py:41]:
```suggestion
                If no CUDA devices are found.
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/accelerators/gpu.py
Comment (ananthsub) [inline pytorch_lightning/accelerators/gpu.py:46]: @carmocca @awaelchli going through #11448, this logic is still replicated across the trainer & each accelerator to support `accelerator="auto"`

What do you think of this instead?
```py
Accelerator(ABC):

@staticmethod
@abstractmethod
def is_available() -> bool:

def setup_environment(self):
    if not self.is_available():
        raise MisconfigurationException("{self.__class__.__qualname__} is not configured to run in this environment")
```

This would function pretty much the same as how `ClusterEnvironment.detect()` works, which means we don't need to duplicate logic across individual accelerators & the accelerator connector.

And the runtime check during trainer execution is automatically applied without being a breaking change that this PR currently introduces.

cc @four4fish

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/accelerators/gpu.py
Comment (justusschock) [inline pytorch_lightning/accelerators/gpu.py:85]:
```suggestion
        return torch.cuda.is_available() and torch.cuda.device_count() > 0
```
 Isn't that redundant? Doesn't `torch.cuda.is_available()` only return `True` if there are also some devices?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/accelerators/gpu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/gpu.py:85]: you're right, this is redundant. device_count internally calls `is_available`

**Review by rohitgr7 — DISMISSED** [review]

File: pytorch_lightning/accelerators/gpu.py
Comment (rohitgr7) [inline pytorch_lightning/accelerators/gpu.py:42]: with DDP it will now fail after `init_dist_connection`

File: pytorch_lightning/accelerators/tpu.py
Comment (rohitgr7) [inline pytorch_lightning/accelerators/tpu.py:52]: for TPU I think it will fail early if no TPU will be available because we call `connect` before `setup_environment`.

File: pytorch_lightning/accelerators/accelerator.py
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:44]: If I dont' have a GPU, this says "GPUAccelerator is not configured to run on this hardware"

This makes it sounds like there's a setting for GPUAccelerator that would make it work, when in reality I don't even have the hardware.

I would suggest rewording to "GPUAccelerator cannot run on this hardware" or something similar

File: pytorch_lightning/accelerators/cpu.py
Comment (carmocca) [inline pytorch_lightning/accelerators/cpu.py:35]: Would a check like this be moved to `is_available`?

File: pytorch_lightning/accelerators/ipu.py
Comment (carmocca) [inline pytorch_lightning/accelerators/ipu.py:38]: You don't need to do this here but I think it would be nicer to move the _IPU_AVAILABLE implementation here and remove this global variable.

Same comment for the other classes

File: tests/plugins/test_amp_plugins.py
Comment (carmocca) [inline tests/plugins/test_amp_plugins.py:48]: I wonder if we could get away with patching the newly introduced staticmethods instead of torch's. No need to do this in this PR

File: pytorch_lightning/accelerators/gpu.py
Comment (awaelchli) [reply pytorch_lightning/accelerators/gpu.py:42]: This is currently a no-op. The super implementation is empty.

**Comment by carmocca** [comment]
Does this close #11799 and #11798?

File: tests/accelerators/test_cpu.py
Comment (awaelchli) [inline tests/accelerators/test_cpu.py:26]:
```suggestion
    assert CPUAccelerator.is_available()
```
otherwise it's tautological :)

File: tests/accelerators/test_ddp.py
Comment (awaelchli) [inline tests/accelerators/test_ddp.py:100]: the order of arguments is incorrect. the decorator closest to the test function corresponds to the first argument, and the one on the very top corresponds to the right-most argument. 

it does not matter here because these are not being used, just as a note.

File: pytorch_lightning/accelerators/accelerator.py
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:43]: I think this should either be called in the init or not called anywhere and we let each accelerator choose.

We cannot be sure the device won't be accessed before `setup_environment`.

File: tests/accelerators/test_ddp.py
Comment (awaelchli) [reply tests/accelerators/test_ddp.py:100]: this applies to other test changes below as well.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/accelerators/ipu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/ipu.py:38]: I agree, I don't think the IPU available and TPU available make sense under imports. That's a good follow up for a separate PR

File: pytorch_lightning/accelerators/accelerator.py
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:43]: I agree, we should first make sure setup_env() is the first method to be called in the run method and omit this check until then. 
Also, I prefer not to have this in init, as instantiating an accelerator should be allowed and running it is a different concern.

File: pytorch_lightning/accelerators/gpu.py
Comment (rohitgr7) [reply pytorch_lightning/accelerators/gpu.py:42]: I am actually pointing to this: https://github.com/PyTorchLightning/pytorch-lightning/blob/8394770d4afa5480f881229b150ac44eaa8c41b0/pytorch_lightning/strategies/ddp.py#L145-L151

File: pytorch_lightning/accelerators/accelerator.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:43]: > We cannot be sure the device won't be accessed before setup_environment. Examples of this are shown in the comments by @rohitgr7

My aim with adding this to `setup_environment` by default is this represents the *latest* time by when the device check should run. This should not preclude earlier possible checks from running

File: pytorch_lightning/accelerators/cpu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/cpu.py:35]: is_available doesn't take an input arguments

accepting a device makes it more like `is_compatible` which is a different intention than a global availability check

File: pytorch_lightning/accelerators/tpu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/tpu.py:52]: For TPU and IPU these checks might feel redundant because there's only a limited set of strategies the accelerator is  paired with. Someone may write a custom TPU training strategy, forget to add that check, but get covered by this one.

File: pytorch_lightning/accelerators/ipu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/ipu.py:38]: @carmocca there's also a whole XLADeviceUtils which probably can be deduped further for TPUs @kaushikb11 https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/xla_device.py

File: pytorch_lightning/accelerators/accelerator.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:43]: @rohitgr7 @awaelchli @carmocca to separate concerns, I can move the check for is_available into a separate PR to unblock this one. Even without this, the rest of the PR is still helpful for the accelerator connector rewrite

**Comment by ananthsub** [comment]
> Does this close #11799 and #11798?

Yes, I will close them after this is merged

File: pytorch_lightning/accelerators/cpu.py
Comment (rohitgr7) [inline pytorch_lightning/accelerators/cpu.py:34]: is this required now?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/accelerators/cpu.py
Comment (ananthsub) [reply pytorch_lightning/accelerators/cpu.py:34]: i added it since it overrides the super class, just in case the super class changes its current no-op implementation to something more meaningful later

**Pull request merged**
